# Clustering Toronto Neighborhoods

First we import the necessary libraries

In [2]:
import requests
!conda install -c anaconda lxml --yes
import lxml.html as lh
import pandas as pd

Solving environment: done


==> WARNING: A newer version of conda exists. <==
  current version: 4.5.11
  latest version: 4.7.12

Please update conda by running

    $ conda update -n base -c defaults conda



## Package Plan ##

  environment location: /home/jupyterlab/conda/envs/python

  added / updated specs: 
    - lxml


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    numpy-base-1.15.4          |   py36h81de0dd_0         4.2 MB  anaconda
    numpy-1.15.4               |   py36h1d66e8a_0          35 KB  anaconda
    certifi-2019.9.11          |           py36_0         154 KB  anaconda
    openssl-1.1.1              |       h7b6447c_0         5.0 MB  anaconda
    lxml-4.3.0                 |   py36hefd8a0e_0         1.5 MB  anaconda
    mkl_fft-1.0.6              |   py36h7dd41cf_0         150 KB  anaconda
    blas-1.0                   |              mkl           6 KB  anaconda
  

Next, we access the URL with the dataset

In [3]:
url='https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
#Then we assign a variable to the the web contents
page = requests.get(url)
#We can then store those contents in a doc variable
doc = lh.fromstring(page.content)
#We then get the contnets within the tr data field
tr_elements = doc.xpath('//tr')

Then we check row width to make sure the table dimensions look accurate

In [4]:
#Check row length of the first 10 rows
[len(T) for T in tr_elements[:10]]

[3, 3, 3, 3, 3, 3, 3, 3, 3, 3]

All are length 3, which is what we would expect, meaning we have grabbed the right table. <br>
Then we take the elements in the first row and make them our headers.

In [5]:
tr_elements = doc.xpath('//tr')
#Create list
col=[]
i=0
#store element and list for each row
for t in tr_elements[0]:
    i+=1
    name=t.text_content()
    print ('%d:"%s"'%(i,name))
    col.append((name, []))
    
    


1:"Postcode"
2:"Borough"
3:"Neighbourhood
"


Next we put the data into a pandas dataframe

In [6]:
#We store the data starting in the second row (after the header)
for j in range(1,len(tr_elements)):
    #T is equal to row j
    T=tr_elements[j]
    
    #If row is greater than 3 columns then don't include as it isnt from the table we need
    if len(T)!=3:
        break
    

        
    #create index i
    i=0
    
    #Run through each row
    for t in T.iterchildren():
        data=t.text_content() 
        #See if there is an empty row
        if i>0:
        #Convert any data that happens to show up numerical as integer
            try:
                data=str(data)
            except:
                pass
        #Append the data to the empty list of the i'th column
        col[i][1].append(data)
        #Increment i for the next column
        i+=1

We double check column length to ensure accuracy

In [7]:
[len(C) for (title,C) in col]

[287, 287, 287]

Next we create the dataframe

In [8]:
Dict={title:column for (title,column) in col}
df=pd.DataFrame(Dict)

We check to make sure the data imported correctly

In [9]:
df.head()

Postcode           Borough     Neighbourhood\n
0      M1A      Not assigned      Not assigned\n
1      M2A      Not assigned      Not assigned\n
2      M3A        North York         Parkwoods\n
3      M4A        North York  Victoria Village\n
4      M5A  Downtown Toronto      Harbourfront\n

The data came in with a newline character in the last column. We need to remove it.

In [10]:
df = df.replace(r'\n',' ', regex=True) 
df.head()

Postcode           Borough    Neighbourhood\n
0      M1A      Not assigned      Not assigned 
1      M2A      Not assigned      Not assigned 
2      M3A        North York         Parkwoods 
3      M4A        North York  Victoria Village 
4      M5A  Downtown Toronto      Harbourfront

Now we need to rename the head without the newline character.

In [11]:
df = df.rename(columns={'Neighbourhood\n': 'Neighbourhood'})
df.head()

Postcode           Borough      Neighbourhood
0      M1A      Not assigned      Not assigned 
1      M2A      Not assigned      Not assigned 
2      M3A        North York         Parkwoods 
3      M4A        North York  Victoria Village 
4      M5A  Downtown Toronto      Harbourfront

Next, we remove the 'not assigned' boroughs.

In [12]:
# Get names of indexes where borough has a value of "not assigned"
indexes = df[ df['Borough'] == 'Not assigned' ].index
# Delete these row indexes from dataFrame
df.drop(indexes , inplace=True)
df.head()

Postcode           Borough      Neighbourhood
2      M3A        North York         Parkwoods 
3      M4A        North York  Victoria Village 
4      M5A  Downtown Toronto      Harbourfront 
5      M6A        North York  Lawrence Heights 
6      M6A        North York    Lawrence Manor

Next we group the neighborhoods of the same postal code. We'll also rename the data frame here in case we need to go back to a point of reference.

In [13]:
df_Toronto =df.groupby(['Postcode','Borough'])['Neighbourhood'].apply(', '.join).reset_index()
df_Toronto.head()

Postcode      Borough                              Neighbourhood
0      M1B  Scarborough                           Rouge , Malvern 
1      M1C  Scarborough  Highland Creek , Rouge Hill , Port Union 
2      M1E  Scarborough       Guildwood , Morningside , West Hill 
3      M1G  Scarborough                                    Woburn 
4      M1H  Scarborough                                 Cedarbrae

Next, we assign the borough name as the neighborhood name when neighborhood is not assigned. 

In [14]:
df_Toronto.loc[df_Toronto['Neighbourhood']=='Not assigned','Neighbourhood'] = df_Toronto.loc[df_Toronto['Neighbourhood']=='Not assigned']['Borough']
df_Toronto.head(10)


Postcode      Borough                                      Neighbourhood
0      M1B  Scarborough                                   Rouge , Malvern 
1      M1C  Scarborough          Highland Creek , Rouge Hill , Port Union 
2      M1E  Scarborough               Guildwood , Morningside , West Hill 
3      M1G  Scarborough                                            Woburn 
4      M1H  Scarborough                                         Cedarbrae 
5      M1J  Scarborough                               Scarborough Village 
6      M1K  Scarborough     East Birchmount Park , Ionview , Kennedy Park 
7      M1L  Scarborough                 Clairlea , Golden Mile , Oakridge 
8      M1M  Scarborough  Cliffcrest , Cliffside , Scarborough Village W...
9      M1N  Scarborough                      Birch Cliff , Cliffside West

Next we import the numpy library to use the shape function to determine the total number of rows and columns in the final table.

In [15]:
import numpy as np

We call the shape function:

In [16]:
print(df_Toronto.shape)

(103, 3)


The table has 103 rows and 3 columns.

Next we use the Longitude Latitude CSV file to get the latitude and longitude of each postal code. We do not use geocoder to automatically get postcodes because it did not function properly upon use.

In [45]:
df_longlat = pd.read_csv (r'http://cocl.us/Geospatial_data')
#rename column so we can do a join
df_ll=df_longlat.rename(columns={'Postal Code': 'Postcode'})

print (df_ll)


    Postcode   Latitude  Longitude
0        M1B  43.806686 -79.194353
1        M1C  43.784535 -79.160497
2        M1E  43.763573 -79.188711
3        M1G  43.770992 -79.216917
4        M1H  43.773136 -79.239476
..       ...        ...        ...
98       M9N  43.706876 -79.518188
99       M9P  43.696319 -79.532242
100      M9R  43.688905 -79.554724
101      M9V  43.739416 -79.588437
102      M9W  43.706748 -79.594054

[103 rows x 3 columns]


Next we merge the two dataframes on postcode, the common column.

In [48]:


df_combined = pd.merge(df_Toronto,
                 df_ll[['Postcode', 'Latitude', 'Longitude']],
                 on='Postcode')
df_combined.head()

Postcode      Borough                              Neighbourhood   Latitude  \
0      M1B  Scarborough                           Rouge , Malvern   43.806686   
1      M1C  Scarborough  Highland Creek , Rouge Hill , Port Union   43.784535   
2      M1E  Scarborough       Guildwood , Morningside , West Hill   43.763573   
3      M1G  Scarborough                                    Woburn   43.770992   
4      M1H  Scarborough                                 Cedarbrae   43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

Importing all packages for map visualization

In [49]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Solving environment: done


==> WARNING: A newer version of conda exists. <==
  current version: 4.5.11
  latest version: 4.7.12

Please update conda by running

    $ conda update -n base -c defaults conda



## Package Plan ##

  environment location: /home/jupyterlab/conda/envs/python

  added / updated specs: 
    - geopy


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    geopy-1.20.0               |             py_0          57 KB  conda-forge
    certifi-2019.11.28         |           py36_0         149 KB  conda-forge
    geographiclib-1.50         |             py_0          34 KB  conda-forge
    ------------------------------------------------------------
                                           Total:         240 KB

The following NEW packages will be INSTALLED:

    geographiclib: 1.50-py_0        conda-forge
    geopy:         1.20.0-py_0      conda-forge

The following pac

Getting Toronto Coordinates

In [50]:
address = 'Toronto, Canada'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.653963, -79.387207.


Now we'll include only Boroughs that contain the word Toronto to define the clustering to the city.

In [88]:
df_combined = df_combined.rename(columns={'Neighbourhood': 'Neighborhood'})
toronto_data = df_combined[df_combined['Borough'].str.contains( 'Toronto')].reset_index(drop=False)

toronto_data.head(20)

index Postcode           Borough  \
0      37      M4E      East Toronto   
1      41      M4K      East Toronto   
2      42      M4L      East Toronto   
3      43      M4M      East Toronto   
4      44      M4N   Central Toronto   
5      45      M4P   Central Toronto   
6      46      M4R   Central Toronto   
7      47      M4S   Central Toronto   
8      48      M4T   Central Toronto   
9      49      M4V   Central Toronto   
10     50      M4W  Downtown Toronto   
11     51      M4X  Downtown Toronto   
12     52      M4Y  Downtown Toronto   
13     53      M5A  Downtown Toronto   
14     54      M5B  Downtown Toronto   
15     55      M5C  Downtown Toronto   
16     56      M5E  Downtown Toronto   
17     57      M5G  Downtown Toronto   
18     58      M5H  Downtown Toronto   
19     59      M5J  Downtown Toronto   

                                         Neighborhood   Latitude  Longitude  
0                                        The Beaches   43.676357 -79.293031  
1                      The Danforth West , Riverdale   43.679557 -79.352188  
2                    The Beaches West , India Bazaar   43.668999 -79.315572  
3                                    Studio District   43.659526 -79.340923  
4                                      Lawrence Park   43.728020 -79.388790  
5                                   Davisville North   43.712751 -79.390197  
6                                 North Toronto West   43.715383 -79.405678  
7                                         Davisville   43.704324 -79.388790  
8                       Moore Park , Summerhill East   43.689574 -79.383160  
9   Deer Park , Forest Hill SE , Rathnelly , South...  43.686412 -79.400049  
10                                          Rosedale   43.679563 -79.377529  
11                      Cabbagetown , St. James Town   43.667967 -79.367675  
12                              Church and Wellesley   43.665860 -79.383160  
13                                      Harbourfront   43.654260 -79.360636  
14                         Ryerson , Garden District   43.657162 -79.378937  
15                                    St. James Town   43.651494 -79.375418  
16                                       Berczy Park   43.644771 -79.373306  
17                                Central Bay Street   43.657952 -79.387383  
18                        Adelaide , King , Richmond   43.650571 -79.384568  
19  Harbourfront East , Toronto Islands , Union St...  43.640816 -79.381752

Next we'll visualize the neighborhoods in Toronto.

In [90]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(toronto_data['Latitude'], toronto_data['Longitude'], toronto_data['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

To cluster the neighborhoods, we'll need to get their features through Foursquare API. I'll define my credentials here.

In [67]:
CLIENT_ID = 'ZTHXBVFLF30SOA4CR21SYOCX50C4O452POKT5Y1PDXL1FHSQ' # your Foursquare ID
CLIENT_SECRET = 'J01LSJTD30T4UMNX24024NG5EWMU3K3RCM1BTDUPHFPAWIKY' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: ZTHXBVFLF30SOA4CR21SYOCX50C4O452POKT5Y1PDXL1FHSQ
CLIENT_SECRET:J01LSJTD30T4UMNX24024NG5EWMU3K3RCM1BTDUPHFPAWIKY


Next we'll get all venue data associated with Toronto Neighborhoods

In [80]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        LIMIT = 100 # limit of number of venues returned by Foursquare API

        radius = 500 # define radius
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [91]:

toronto_venues = getNearbyVenues(names=toronto_data['Neighborhood'],
                                   latitudes=toronto_data['Latitude'],
                                   longitudes=toronto_data['Longitude'])


The Beaches 
The Danforth West , Riverdale 
The Beaches West , India Bazaar 
Studio District 
Lawrence Park 
Davisville North 
North Toronto West 
Davisville 
Moore Park , Summerhill East 
Deer Park , Forest Hill SE , Rathnelly , South Hill , Summerhill West 
Rosedale 
Cabbagetown , St. James Town 
Church and Wellesley 
Harbourfront 
Ryerson , Garden District 
St. James Town 
Berczy Park 
Central Bay Street 
Adelaide , King , Richmond 
Harbourfront East , Toronto Islands , Union Station 
Design Exchange , Toronto Dominion Centre 
Commerce Court , Victoria Hotel 
Roselawn 
Forest Hill North , Forest Hill West 
The Annex , North Midtown , Yorkville 
Harbord , University of Toronto 
Chinatown , Grange Park , Kensington Market 
CN Tower , Bathurst Quay , Island airport , Harbourfront West , King and Spadina , Railway Lands , South Niagara 
Stn A PO Boxes 25 The Esplanade 
First Canadian Place , Underground city 
Christie 
Dovercourt Village , Dufferin 
Little Portugal , Trinity 
Brockton ,

Next we'll check the resulting dataframe

In [92]:
print(toronto_venues.shape)
toronto_venues.head()

(1685, 7)


Neighborhood  Neighborhood Latitude  \
0                    The Beaches               43.676357   
1                    The Beaches               43.676357   
2                    The Beaches               43.676357   
3                    The Beaches               43.676357   
4  The Danforth West , Riverdale               43.679557   

   Neighborhood Longitude                               Venue  Venue Latitude  \
0              -79.293031                   Glen Manor Ravine       43.676821   
1              -79.293031  The Big Carrot Natural Food Market       43.678879   
2              -79.293031                 Grover Pub and Grub       43.679181   
3              -79.293031                       Upper Beaches       43.680563   
4              -79.352188                            Pantheon       43.677621   

   Venue Longitude     Venue Category  
0       -79.293942              Trail  
1       -79.297734  Health Food Store  
2       -79.297215                Pub  
3       -79.292869       Neighborhood  
4       -79.351434   Greek Restaurant

Let's check the number of venues per neighborhood

In [93]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Adelaide , King , Richmond                                            100   
Berczy Park                                                            57   
Brockton , Exhibition Place , Parkdale Village                         21   
Business Reply Mail Processing Centre 969 Eastern                      16   
CN Tower , Bathurst Quay , Island airport , Har...                     16   
Cabbagetown , St. James Town                                           46   
Central Bay Street                                                     85   
Chinatown , Grange Park , Kensington Market                            93   
Christie                                                               17   
Church and Wellesley                                                   83   
Commerce Court , Victoria Hotel                                       100   
Davisville                                                             37   
Davisville North                                                        9   
Deer Park , Forest Hill SE , Rathnelly , South ...                     15   
Design Exchange , Toronto Dominion Centre                             100   
Dovercourt Village , Dufferin                                          14   
First Canadian Place , Underground city                               100   
Forest Hill North , Forest Hill West                                    4   
Harbord , University of Toronto                                        35   
Harbourfront                                                           49   
Harbourfront East , Toronto Islands , Union Sta...                    100   
High Park , The Junction South                                         25   
Lawrence Park                                                           3   
Little Portugal , Trinity                                              62   
Moore Park , Summerhill East                                            2   
North Toronto West                                                     19   
Parkdale , Roncesvalles                                                14   
Rosedale                                                                4   
Roselawn                                                                1   
Runnymede , Swansea                                                    36   
Ryerson , Garden District                                             100   
St. James Town                                                        100   
Stn A PO Boxes 25 The Esplanade                                        98   
Studio District                                                        37   
The Annex , North Midtown , Yorkville                                  20   
The Beaches                                                             4   
The Beaches West , India Bazaar                                        20   
The Danforth West , Riverdale                                          43   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Adelaide , King , Richmond                                             100   
Berczy Park                                                             57   
Brockton , Exhibition Place , Parkdale Village                          21   
Business Reply Mail Processing Centre 969 Eastern                       16   
CN Tower , Bathurst Quay , Island airport , Har...                      16   
Cabbagetown , St. James Town                                            46   
Central Bay Street                                                      85   
Chinatown , Grange Park , Kensington Market                             93   
Christie                                                                17   
Church and Wellesley                                                    83   
Commerce Court , Victoria Hotel    

Now let's check the number of categories per venue

In [84]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 233 uniques categories.


Now we'll use one hot coding to analyze the neighborhoods

In [94]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Yoga Studio  Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0            0                  0        0                   0             0   
1            0                  0        0                   0             0   
2            0                  0        0                   0             0   
3            0                  0        0                   0             0   
4            0                  0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  Argentinian Restaurant  Art Gallery  \
0             0         0                       0            0   
1             0         0                       0            0   
2             0         0                       0            0   
3             0         0                       0            0   
4             0         0                       0            0   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Auto Workshop  \
0                    0                 0                   0              0   
1                    0                 0                   0              0   
2                    0                 0                   0              0   
3                    0                 0                   0              0   
4                    0                 0                   0              0   

   BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  Baseball Stadium  \
0          0           0           0       0     0    0                 0   
1          0           0           0       0     0    0                 0   
2          0           0           0       0     0    0                 0   
3          0           0           0       0     0    0                 0   
4          0           0           0       0     0    0                 0   

   Basketball Stadium  Beach  Bed & Breakfast  Beer Bar  Beer Store  \
0                   0      0                0         0           0   
1                   0      0                0         0           0   
2                   0      0                0         0           0   
3                   0      0                0         0           0   
4                   0      0                0         0           0   

   Belgian Restaurant  Bistro  Board Shop  Boat or Ferry  Bookstore  Boutique  \
0                   0       0           0              0          0         0   
1                   0       0           0              0          0         0   
2                   0       0           0              0          0         0   
3                   0       0           0              0          0         0   
4                   0       0           0              0          0         0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bubble Tea Shop  Building  \
0                     0               0        0                0         0   
1                     0               0        0                0         0   
2                     0               0        0                0         0   
3                     0               0        0                0         0   
4                     0               0        0                0         0   

   Burger Joint  Burrito Place  Bus Line  Butcher  Café  \
0             0              0         0        0     0   
1             0              0         0        0     0   
2             0              0         0        0     0   
3             0              0         0        0     0   
4             0              0         0

Check dataframe size

In [95]:
toronto_onehot.shape

(1685, 233)

Group rows by mean frequency

In [96]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Yoga Studio  \
0                         Adelaide , King , Richmond      0.000000   
1                                        Berczy Park      0.000000   
2     Brockton , Exhibition Place , Parkdale Village      0.000000   
3   Business Reply Mail Processing Centre 969 East...     0.000000   
4   CN Tower , Bathurst Quay , Island airport , Ha...     0.000000   
5                       Cabbagetown , St. James Town      0.000000   
6                                 Central Bay Street      0.011765   
7        Chinatown , Grange Park , Kensington Market      0.000000   
8                                           Christie      0.000000   
9                               Church and Wellesley      0.012048   
10                   Commerce Court , Victoria Hotel      0.000000   
11                                        Davisville      0.000000   
12                                  Davisville North      0.000000   
13  Deer Park , Forest Hill SE , Rathnelly , South...     0.000000   
14         Design Exchange , Toronto Dominion Centre      0.000000   
15                     Dovercourt Village , Dufferin      0.000000   
16           First Canadian Place , Underground city      0.000000   
17              Forest Hill North , Forest Hill West      0.000000   
18                   Harbord , University of Toronto      0.000000   
19                                      Harbourfront      0.020408   
20  Harbourfront East , Toronto Islands , Union St...     0.000000   
21                    High Park , The Junction South      0.000000   
22                                     Lawrence Park      0.000000   
23                         Little Portugal , Trinity      0.016129   
24                      Moore Park , Summerhill East      0.000000   
25                                North Toronto West      0.052632   
26                           Parkdale , Roncesvalles      0.000000   
27                                          Rosedale      0.000000   
28                                          Roselawn      0.000000   
29                               Runnymede , Swansea      0.000000   
30                         Ryerson , Garden District      0.000000   
31                                    St. James Town      0.000000   
32                   Stn A PO Boxes 25 The Esplanade      0.000000   
33                                   Studio District      0.027027   
34             The Annex , North Midtown , Yorkville      0.000000   
35                                       The Beaches      0.000000   
36                   The Beaches West , India Bazaar      0.000000   
37                     The Danforth West , Riverdale      0.023256   

    Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0            0.000000   0.0000              0.0000        0.0000   
1            0.000000   0.0000              0.0000        0.0000   
2            0.000000   0.0000              0.0000        0.0000   
3            0.000000   0.0000              0.0000        0.0000   
4            0.000000   0.0625              0.0625        0.0625   
5            0.000000   0.0000              0.0000        0.0000   
6            0.000000   0.0000              0.0000        0.0000   
7            0.000000   0.0000              0.0000        0.0000   
8            0.000000   0.0000              0.0000        0.0000   
9            0.012048   0.0000              0.0000        0.0000   
10           0.000000   0.0000              0.0000        0.0000   
11           0.000000   0.0000              0.0000        0.0000   
12           0.000000   0.0000              0.0000        0.0000   
13           0.000000   0.0000              0.0000        0.0000   
14           0.000000   0.0000              0.0000        0.0000   
15           0.000000   0.0000              0.0000        0.0000   
16           0.000000   0.0000              0.0000        0.0000   
17           0.000000   0.0000              0.0000        0.0000   
18           0.000

Confirm size

In [98]:
toronto_grouped.shape

(38, 233)

Let's examine each neighborhood with the top five vanues there.

In [99]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Adelaide , King , Richmond ----
             venue  freq
0      Coffee Shop  0.07
1             Café  0.05
2       Steakhouse  0.04
3              Bar  0.04
4  Thai Restaurant  0.04


----Berczy Park ----
            venue  freq
0     Coffee Shop  0.07
1          Bakery  0.05
2     Cheese Shop  0.04
3  Farmers Market  0.04
4        Beer Bar  0.04


----Brockton , Exhibition Place , Parkdale Village ----
                venue  freq
0      Breakfast Spot  0.10
1                Café  0.10
2         Coffee Shop  0.10
3  Italian Restaurant  0.05
4       Burrito Place  0.05


----Business Reply Mail Processing Centre 969 Eastern ----
                  venue  freq
0           Pizza Place  0.06
1         Auto Workshop  0.06
2               Brewery  0.06
3  Fast Food Restaurant  0.06
4        Farmers Market  0.06


----CN Tower , Bathurst Quay , Island airport , Harbourfront West , King and Spadina , Railway Lands , South Niagara ----
              venue  freq
0   Airport Service  0.12
1  A

Now we need to place that data into a dataframe

In [100]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [119]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                        Adelaide , King , Richmond            Coffee Shop   
1                                       Berczy Park            Coffee Shop   
2    Brockton , Exhibition Place , Parkdale Village         Breakfast Spot   
3  Business Reply Mail Processing Centre 969 East...            Smoke Shop   
4  CN Tower , Bathurst Quay , Island airport , Ha...       Airport Service   

  2nd Most Common Venue 3rd Most Common Venue  4th Most Common Venue  \
0                  Café            Steakhouse        Thai Restaurant   
1                Bakery        Farmers Market               Beer Bar   
2           Coffee Shop                  Café  Performing Arts Venue   
3                Garden            Skate Park                Brewery   
4      Airport Terminal       Harbor / Marina            Coffee Shop   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                   Bar          Burger Joint                Bakery   
1          Cocktail Bar    Seafood Restaurant            Steakhouse   
2          Climbing Gym                   Bar        Sandwich Place   
3        Farmers Market  Fast Food Restaurant                   Spa   
4              Boutique         Boat or Ferry   Rental Car Location   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0      Asian Restaurant      Sushi Restaurant    American Restaurant  
1           Cheese Shop                  Café    Japanese Restaurant  
2         Burrito Place            Restaurant              Pet Store  
3            Restaurant         Burrito Place          Auto Workshop  
4                   Bar                 Plane       Sculpture Garden

Finally, we will use K means clustering the cluster the neighborhoods into 4 clusters

In [120]:
# set number of clusters
kclusters = 4


toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0], dtype=int32)

Now we add the clusters to the dataframe

In [121]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = toronto_data

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head() # check the last columns!

index Postcode          Borough                      Neighborhood  \
0     37      M4E     East Toronto                      The Beaches    
1     41      M4K     East Toronto    The Danforth West , Riverdale    
2     42      M4L     East Toronto  The Beaches West , India Bazaar    
3     43      M4M     East Toronto                  Studio District    
4     44      M4N  Central Toronto                    Lawrence Park    

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.676357 -79.293031               3     Health Food Store   
1  43.679557 -79.352188               0      Greek Restaurant   
2  43.668999 -79.315572               0                  Park   
3  43.659526 -79.340923               0                  Café   
4  43.728020 -79.388790               3                  Park   

  2nd Most Common Venue 3rd Most Common Venue    4th Most Common Venue  \
0                 Trail                   Pub              Wings Joint   
1           Coffee Shop        Ice Cream Shop       Italian Restaurant   
2        Ice Cream Shop            Board Shop           Sandwich Place   
3           Coffee Shop    Italian Restaurant      American Restaurant   
4           Swim School              Bus Line  Comfort Food Restaurant   

  5th Most Common Venue   6th Most Common Venue 7th Most Common Venue  \
0      Department Store    Ethiopian Restaurant     Electronics Store   
1             Bookstore  Furniture / Home Store            Restaurant   
2               Brewery            Burger Joint  Fast Food Restaurant   
3                Bakery          Ice Cream Shop                  Park   
4  Colombian Restaurant             Event Space  Ethiopian Restaurant   

         8th Most Common Venue        9th Most Common Venue  \
0  Eastern European Restaurant          Dumpling Restaurant   
1               Cosmetics Shop                      Brewery   
2                Burrito Place           Italian Restaurant   
3                    Bookstore           Seafood Restaurant   
4            Electronics Store  Eastern European Restaurant   

  10th Most Common Venue  
0             Donut Shop  
1        Bubble Tea Shop  
2      Fish & Chips Shop  
3         Sandwich Place  
4    Dumpling Restaurant

To Finish, we will view our clusters.

In [122]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters